### PR TITLE
hal/sparcv8leon: implement high precision timer

### PR DESCRIPTION
**Pull request by lukileczo**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: `sparcv8leon-gr716-mimas`, `sparcv8leon-generic-qemu`, `sparcv8leon-gr740-mini`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
https://github.com/phoenix-rtos/phoenix-rtos-project/pull/1228
https://github.com/phoenix-rtos/plo/pull/369
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/sparcv8leon/gaisler/timer.c [hunk 1]
```diff
@@ -40,31 +40,49 @@
...
-		hal_cpuDataStoreBarrier();
+		if (timer == TIMER_TIMEBASE) {
+			++timer_common.jiffies;
+		}
+		timer_clearIrq(timer);
 	}
 
+	hal_spinlockClear(&timer_common.sp, &sc);
+
 	return 0;
```
Comment (agkaminski) [inline hal/sparcv8leon/gaisler/timer.c:86]: Shouldn't we return 1 on TIMER_WAKEUP (to trigger reschedule)?

File: hal/sparcv8leon/gaisler/timer.c [hunk 2]
```diff
@@ -75,40 +93,64 @@ static inline void timer_setReloadValue(int timer, u32 val)
...
+
+	hal_spinlockSet(&timer_common.sp, &sc);
+
+	/* Disable timer */
+	*(timer_common.timer0_base + GPT_TCTRL(TIMER_WAKEUP)) = 0;
+	timer_clearIrq(TIMER_WAKEUP);
+
+	/* Configure one shot timer */
+	timer_setReloadValue(TIMER_WAKEUP, waitUs);
+	*(timer_common.timer0_base + GPT_TCTRL(TIMER_WAKEUP)) = TIMER_ENABLE | TIMER_INT_ENABLE | TIMER_LOAD;
```
Comment (agkaminski) [inline hal/sparcv8leon/gaisler/timer.c:144]: Just making sure - `TIMER_LOAD` flag cause reload value to be transferred to the counter register, so the timer starts with the reload value, not the old value of the counter?

File: hal/sparcv8leon/hal.c [hunk 3]
```diff
@@ -97,5 +97,5 @@ void _hal_init(void)
 	_hal_platformInit();
 	_hal_cpuInit();
 	_hal_consoleInit();
-	_hal_timerInit(SYSTICK_INTERVAL);
+	_hal_timerInit(0xffffffffu);
```
Comment (agkaminski) [inline hal/sparcv8leon/hal.c:100]: Interval is still `SYSTICK_INTERVAL` - in HP timer case it's value still should be used as a maximum interval period for the wakeup timer

File: hal/sparcv8leon/gaisler/timer.c [hunk 4]
```diff
@@ -40,31 +40,49 @@
...
+	hal_cpuDataStoreBarrier();
+	*(timer_common.timer0_base + GPT_TCTRL(timer)) &= ~TIMER_INT_PENDING;
+	hal_cpuDataStoreBarrier();
+}
+
+
 static int _timer_irqHandler(unsigned int irq, cpu_context_t *ctx, void *data)
 {
-	volatile u32 st = *(timer_common.timer0_base + GPT_TCTRL(TIMER_DEFAULT)) & TIMER_INT_PENDING;
+	volatile u32 st;
```
Comment (agkaminski) [inline hal/sparcv8leon/gaisler/timer.c:69]: Why is this volatile?

File: hal/sparcv8leon/hal.c (hunk 3 above)
Comment (lukileczo) [reply hal/sparcv8leon/hal.c:100]: The only place where `hal_timerSetWakeup` is called is in `_threads_updateWakeup` and there is a check for the wakeup value: https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/e2a6b1f764dc01ed672afc2bf539a1fd3e147f1a/proc/threads.c#L418
So IMO there's no need to check for interval in timer code

**Review by agkaminski — APPROVED** [review]

**Pull request merged**